website/sales/models/order.py

```diff
@@ -115,16 +115,17 @@ def save(
         self, force_insert=False, force_update=False, using=None, update_fields=None
     ):
         if self.shift.locked:
-            return
+            raise ValueError("The shift this order belongs to is locked.")
         if self.shift.start > timezone.now():
-            return
+            raise ValueError("The shift hasn't started yet.")
         if (
             self.payment
-            and float(sum(self.order_items.values_list("total", flat=True)))
-            - (self.discount or 0)
-            != self.payment.amount
+            and self.subtotal - Decimal(self.discount or 0) != self.payment.amount
         ):
-            return
+            # We cannot use self.total_amount as it is a requires a database query and hence will not use any updated values
```

Comment:
But isn't subtotal also using a database query?

Reply:
Yes, but that query only uses other objects (namely OrderItems). The problem with `total_amount` is that it uses the `discount` field of the same object that we're trying to save here (and so it will use the old value that is still in the database whereas we want to compute the new value).